### PR TITLE
Packer: Support building 1.12.0 package

### LINKED ISSUE #178: Handle CSS dependencies from JS comments
As Jörn stated on https://github.com/jquery/jquery-ui/pull/1029#issuecomment-27757535, handle CSS dependencies as discussed.

> We've discussed this at the team meeting in Amsterdam. None of the solutions that load CSS through AMD are in a stage where we're willing to commit to them. In other words, specifying CSS dependencies through AMD is far from a 'solved problem', so we're not going to adopt that approach until it works.

> Since we want to share the download builder between UI and Mobile as soon as possible, we'll adopt Mobile's approach of specifying CSS dependencies in comments in JS files. We'll port the implementation from their build (or download builder). Mobile used this long enough and never ran into any major problems. Once a better solution emerges, we can delete those comments and adopt the new solution.

The jQuery Mobile download builder already [implements this](https://github.com/gseguin/node-amd-builder/blob/master/server.js#L449).

### LINKED ISSUE #255: Super simplify UI 1.12.0 Builder
jQuery UI 1.12 has adopted the Mobile way to declare its CSS dependencies using JS comments. Given this step, we are able to super generalize builder. For example, the following JSON could be used as a configuration-guide on how to build the package.

Maintenance/changes for the package content could ideally be adjusted by fixing the configuration JSON instead of changing the download builder (the builder code).

```js
{
  "external/jquery.js": {
    "bower": {
      "endpoint": "jquery@<version>",
      "file": "dist/jquery.js"
    }
  },
  "images": {
    "themeroller-images": {
      "vars": "{{vars}}"
    }
  },
  "index.html": {
    "template": xxx
  },
  "jquery-ui.css": {
    "concat": [
      "bundle://jquery-ui.structure.css",
      "bundle://jquery-ui.theme.css"
    ]
  },
  "jquery-ui.js": {
    "amd": {
      "include": "{{components}}"
    }
  },
  "jquery-ui.structure.css": {
    "jquery-css": {
      "selector": "structure",
      "include": "{{components}}"
    }
  },
  "jquery-ui.theme.css": {
    "themeroller": {
      "base": {
        "jquery-css": {
          "selector": "theme",
          "include": "{{components}}"
        }
      },
      "vars": "{{vars}}"
    }
  },
  "jquery-ui.min.css": {
    "sqwish": [
      "bundle://jquery-ui.css"
    ]
  },
  "jquery-ui.min.js": {
    "uglify": [
      "bundle://jquery-ui.css"
    ]
  },
  "jquery-ui.structure.min.css": {
    "sqwish": [
      "bundle://jquery-ui.structure.css"
    ]
  },
  "jquery-ui.theme.min.css": {
    "sqwish": [
      "bundle://jquery-ui.theme.css"
    ]
  },
}
```

The idea is having a builder API similar to `Builder( configuration, runtime );`. For example:

```
Builder( configurationAbove, {
  components: [ "core", "widget", ... ]
  vars: {
    ffDefault: "Arial,Helvetica,sans-serif",
    fsDefault: "1em",
    ...
  }
});
```

- [ ] Need to polish this idea.

### PR DESCRIPTION
**Pull request by rxaviers**
- Handle CSS dependencies from JS comments (based on jQuery Mobile);
- Use archive-packager, builder-amd and builder-jquery-css to build UI 1.12.0
  download package;

Fixes #178
Fixes #255
Ref https://github.com/jquery/jquery-ui/pull/1440

TODO:
- [x] [node-packager](https://github.com/rxaviers/node-packager)
- [x] [builder-jquery-css](https://github.com/rxaviers/builder-jquery-css)
  - [x] [builder-amd-css](https://github.com/rxaviers/builder-amd-css)
    - [x] [requirejs-memfiles](https://github.com/rxaviers/requirejs-memfiles)
- [x] [builder-amd](https://github.com/rxaviers/builder-amd)
  - [x] [requirejs-memfiles](https://github.com/rxaviers/requirejs-memfiles)
- [x] [jquery-ui-themeroller](https://github.com/rxaviers/jquery-ui-themeroller)
- [x] tests
- [x] cache 1.12 for improved performance (should be tracked separately)
- SKIP change frontend: dont allow downloading no components selection, remove "No theme" from theme options.

### CODE REVIEW

**Comment by jzaefferer** [comment]
Let me know when this is ready for testing and review.

**Comment by rxaviers** [comment]
Sure, I will.

**Comment by jzaefferer** [comment]
Btw. I think the approach outlined here and in the ticket is sane. I don't know how feasible it will be to move the file to the UI repo, similar to `release.js`. That may need some more abstraction to narrow it down more. And I think to figure out the right abstraction we'd have to also implement the same logic for jQuery Mobile.

File: download.js [hunk 1]
```diff
@@ -101,37 +102,66 @@ Frontend.prototype = {
...
+				Package = require( "./lib/package-1-12" );
+				pkg = new Packager( jqueryUi.files().cache, Package, {
+					components: components,
+					themeVars: themeVars,
+					scope: fields.scope
+				});
+				response.setHeader( "Content-Type", "application/zip" );
+				response.setHeader( "Content-Disposition", "attachment; filename=" + pkg.pkg.zipFilename ); // FIXME
+				pkg.toZip( response, {
+					basedir: pkg.pkg.zipBasedir // FIXME
```
Comment (rxaviers) [inline download.js]: Note to self: find a better name for `pkg.pkg`.

**Comment by rxaviers** [comment]
> Btw. I think the approach outlined here and in the ticket is sane. I don't know how feasible it will be to move the file to the UI repo, similar to release.js. That may need some more abstraction to narrow it down more. And I think to figure out the right abstraction we'd have to also implement the same logic for jQuery Mobile.

Great and ok on waiting to see how this logic will look like for Mobile first before moving it into the respective repos.

**Comment by arschmitz** [comment]
Once this is ready for mobile let me know and we can start the implementation process there to see what this is going to look like and what sort of abstraction will make sense.

**Comment by rxaviers** [comment]
PR updated...

There are some TODO items still. But, this code is functional and is ready for its first review.

Important changes... (a) Previously, it was possible to select no components in order to download a package with the theme only (not exactly the theme only, but kind of). Now, given the fact the theme.css requirement is deduced from the JS components, downloading a package with no components makes no sense and it won't work (similar to jQuery Mobile, which doesn't allow an empty list to be downloaded). (b) Previously, it was possible to select "No theme" in order to download a package with the structure CSS only (not exactly the structure CSS only, but kind of). Now, again given the fact the theme.css requirement is deduced from the JS components, the component selection will tell whether or not theme.css is required. For example, by downloading effects only, a download package won't include theme CSSes. Although, downloading any other component that requires theme.css will include theme CSSes.

**Comment by arschmitz** [comment]
> Previously, it was possible to select "No theme" in order to download a package with the structure CSS only (not exactly the structure CSS only, but kind of)

This is one think I would like to see changed back. I think its important for both projects to be able to exclude the theme file, to make it easier to include their own. While this is not something that is commonly done it does not seem. It is something we have always tried to encourage.

**Comment by rxaviers** [comment]
> This is one think I would like to see changed back. I think its important for both projects to be able to exclude the theme file, to make it easier to include their own. While this is not something that is commonly done it does not seem. It is something we have always tried to encourage.

Any problem with the below?

```diff
- <link type="text/css" rel="stylesheet" href="jquery-ui-1.12.0.custom/jquery-ui.theme.css" />
+ <link type="text/css" rel="stylesheet" href="my.theme.css" />
```

**Comment by rxaviers** [comment]
> Once this is ready for mobile let me know and we can start the implementation process there to see what this is going to look like and what sort of abstraction will make sense.

Ok. If you want to take a quick look, jQuery Mobile package definition will look something like this: https://gist.github.com/rxaviers/d60e586ccae7b3b6ac67

**Comment by arschmitz** [comment]
@rxaviers i'm not really sure what your trying to show. Thats just a diff of a file name changing in a link tag? How does that help someone exclude the theme file? `theme.css` should be optional because its not a true dependency. The widgets operate just fine without it. They just don't look as pretty. This gives you the ability to apply your own theming, with out the need of a ton of specificity work arounds, or manually editing the file ( not at all maintainable ).

**Comment by rxaviers** [comment]
> @rxaviers i'm not really sure what your trying to show. Thats just a diff of a file name changing in a link tag? How does that help someone exclude the theme file? theme.css should be optional because its not a true dependency. The widgets operate just fine without it. They just don't look as pretty. This gives you the ability to apply your own theming, with out the need of a ton of specificity work arounds, or manually editing the file ( not at all maintainable ). 

My diff shows a developer HTML, where instead of using the generated theme.css, developer uses his own my.theme.css.

> This gives you the ability to apply your own theming, with out the need of a ton of specificity work arounds, or manually editing the file ( not at all maintainable ). 

Which ones? the generated `jquery-ui.structure.css` is free from any theme styles. Am I missing anything?

**Comment by arschmitz** [comment]
@rxaviers maybe i'm misunderstanding. You said
>Previously, it was possible to select "No theme" in order to download a package with the structure CSS only.

So if this was only previously possible, how would you get the custom` jquery-ui.structure.css` now, if you can't exclude the theme?

**Comment by rxaviers** [comment]
> So if this was only previously possible, how would you get the customjquery-ui.structure.css now, if you can't exclude the theme?

The "No theme" was created at a `jquery-ui.structure.css` pre-era. Now, we have `jquery-ui.structure.css` in place just like mobile.

**Comment by rxaviers** [comment]
I've edited my previous comment

**Comment by arschmitz** [comment]
@rxaviers Understood yes thats fine then.

**Comment by rxaviers** [comment]
On today's UI meeting, the decision was that drop `theme.css` as a hard-dependency for all JS components (more details [here](https://github.com/jquery/jquery-ui/pull/1440#issuecomment-73933622)). As a consequence, there's no more relationship between JS components and a theme selection. Therefore:

Selecting at least one component will generate:
- external/jquery/jquery.js
- index.html
- jquery-ui.js
- jquery-ui.min.js
- jquery-ui.structure.css
- jquery-ui.structure.min.css

Selecting a theme will generate:
- images
- jquery-ui.theme.css
- jquery-ui.theme.min.css

If both at least one component and a theme were selected, it will generate:
- jquery-ui.css
- jquery-ui.min.css

@jzaefferer @scottgonzalez @arschmitz please could you review?

**Comment by scottgonzalez** [comment]
I would have expected `jquery-ui.css` and `jquery-ui.min.css` as long as any CSS exists, but I'm fine either way.

**Comment by arschmitz** [comment]
> I would have expected jquery-ui.css and jquery-ui.min.css as long as any CSS exists, but I'm fine either way.

I agree here that would have been my expectation as well but i'm fine either way.

**Comment by rxaviers** [comment]
@scottgonzalez and @arschmitz, no problem. So, I've edited my comment with:

```diff
- If both (a) at least one component and (b) a theme were selected, it will generate:
+ If either (a) a component or (b) a theme was selected, it will generate:
```

**Comment by jzaefferer** [comment]
I've added an "also" to "will also generate", since that is in addition to the others. Assuming that's correct, this looks good to me.

**Comment by rxaviers** [comment]
> will also generate

:+1:

**Comment by jzaefferer** [comment]
@rxaviers what's the status of this? Do you need help or input?

**Comment by rxaviers** [comment]
This PR has been updated according to https://github.com/jquery/download.jqueryui.com/pull/256#issuecomment-73939576.

Feel free to test it. For the test, it's sufficient to follow what's in [Getting Started](https://github.com/jquery/download.jqueryui.com/#getting-started) section only.

**Comment by jzaefferer** [comment]
Tested this PR standalone (outside of jqueryui.com):

* The clientside dependencies are broken, for example, while everything is unchecked, checking accordion or autocomplete won't check their dependencies. Their `data-dependencies` attribute has no value.
* When selecting just Draggable (dependencies get selected), the downloaded `index.html` has a layout issue with the "Highlight / Error" header, which floats next to the icons grid
* Looking at ThemeRoller with the default base theme selected, the "Overlay and Shadow Classes" demo is broken (it works fine in the downloaded demo file), here the solid white background of the overlay is missing, rendering the content unreadable. I guess this applies until 1.12 is stable?
* Also on TR: The "Open Dialog" has a tiny font (same problem on the downloaded demo file)

**Comment by rxaviers** [comment]
@jzaefferer, thanks for these catches. Although they probably should be fixed before the UI 1.12 release, none has to do with this PR. So, I suggest we handle them in separate PRs. Details below.

> * The clientside dependencies are broken, for example, while everything is unchecked, checking accordion or autocomplete won't check their dependencies. Their `data-dependencies` attribute has no value.

I've created #271 with details. Could you give your input there please?

> * When selecting just Draggable (dependencies get selected), the downloaded `index.html` has a layout issue with the "Highlight / Error" header, which floats next to the icons grid
> * Looking at ThemeRoller with the default base theme selected, the "Overlay and Shadow Classes" demo is broken (it works fine in the downloaded demo file), here the solid white background of the overlay is missing, rendering the content unreadable. I guess this applies until 1.12 is stable?
> * Also on TR: The "Open Dialog" has a tiny font (same problem on the downloaded demo file)

On #248, the `index.html` (the downloaded demo) have been updated to reflect the changes of jquery/jquery-ui#1384.

On #251, the content for the ThemeRoller has been updated to reflect whatever was in `index.html` (the downloaded demo) for easier maintenance and consistency.

Could you submit a PR fixing what you think should be fixed?

> I guess this applies until 1.12 is stable?

ThemeRoller [loads jQuery UI from CDN](https://github.com/jquery/download.jqueryui.com/blob/6ae5a1e4a214fdca9b4880b5ac8b462c77ec6966/template/themeroller/index.html#L19) using the version defined as stable in [config.json](https://github.com/jquery/download.jqueryui.com/blob/5935d4f54e072c61551e21964d747490a467643e/config.json#L7). So either change the template to use a local package or keep in mind you're testing a different UI version. :)

**Comment by rxaviers** [comment]
Cache support added.

**Comment by jzaefferer** [comment]
I've created #273 for the remaining issues. With #271 fixed and cache support in place, this can land.

**Comment by rxaviers** [comment]
Excellent and thanks. Landed.

**Comment by rxaviers** [comment]
Closed by eda7d35b545950e14c5cb9073b9c98df460ee201

**Pull request closed without merging**